File: 7/q1.hs
import Data.Char

check :: Int -> [Int] -> Bool
check tar params = check' tar params 0
    where
        check' tar [] i     = i==tar
        check' tar (t:ts) i =
            i <= tar &&
            check' tar ts (i+t) ||
            check' tar ts (i*t)

splitOn :: Eq a => a -> [a] -> [[a]]
splitOn c ss = split' ss c []
    where
        split' [] _ sb     = [sb]
        split' (s:ss) c sb =
            if c==s
                then sb : split' ss c []
                else split' ss c (sb ++ [s])

main :: IO ()
main = do
    input <- readFile "input.txt"
    let result = sum $ map ((\z -> if check (head z) (tail z) then head z else 0) . (map (\y -> read y :: Int) . splitOn ' ' . filter (\x -> isDigit x || isSpace x))) (splitOn '\n' input)
    print result

File: 2024/5/q1.py
def c(ls:list, m:list):
    ls = list(map(int, ls.split(',')))
    traversed = []
    for e in ls:
        for t in traversed:
            if t in m[e]:
                return 0
        traversed.append(e)
    return ls[len(ls)//2]

f = open("input.txt", "r").read().split("\n\n")
#f = open("test1.txt", "r").read().split("\n\n")
specs = f[0].split('\n')
xs = f[1].split('\n')


a = [[] for i in range(100)]

for l in specs:
    l = l.split('|')
    k = int(l[0])
    e = int(l[1])
    for i in range(100):
        if i==k:
            a[i] += [e]


print(sum([c(x, a) for x in xs]))